### ISSUE TITLE
how to build go-sqlite3 to include the json feature ?

### ISSUE THREAD

**Issue by jiehuaou** [mattn/go-sqlite3#710]
Here is my install procedure:

1) go get github.com/mattn/go-sqlite3
2) go build -tags json1 github.com/mattn/go-sqlite3
3) go install github.com/mattn/go-sqlite3

after build and install , I run my code like this:

// open database
	database, _ := sql.Open("sqlite3", "file:test.db?cache=shared&mode=memory")
	statement, _ := database.Prepare("CREATE TABLE IF NOT EXISTS people (id INTEGER PRIMARY KEY, key TEXT, value TEXT, valueJson)")
	statement.Exec()

``
	// INSERT data
	statement, _ = database.Prepare("INSERT INTO people (key, valueJson) VALUES (?, ?)")
	statement.Exec("Nic", `{"a":1, "b": "Raboy"}`)
	statement.Exec("Hello", `{"a":2, "b": "World"}`)

	sqlStr := `SELECT id, key, valueJson FROM people where json_extract(valueJson, "$.a")=2`

	rows, err := database.Query(sqlStr)
	if err != nil {
		fmt.Println(err)
	}

`
then I got this error
**no such function: json_extract
panic: runtime error: invalid memory address or nil pointer dereference**

what did I miss ?

**Comment by rittneje** [comment]
> after build and install , I run my code like this

That's not how Go works. There is no need to install dependencies in advance, as Go will always compile your application in full. I'm assuming that when you compiled your actual application (via `go run`, `go build` or `go install`), you did not specify the `json` build tag. Consequently, your application does not include the feature in question, because the version of this library that was actually compiled in was not told to include it.

**Comment by jiehuaou** [comment]
I have fixed it by adding this line in sqlite3.go
`
#cgo CFLAGS: -DSQLITE_ENABLE_JSON1
`
then rebuild and install,

**Comment by glebarez** [comment]
same issue
your solution does seem like a hammer-hit.
why isn't normal way working ?

**Comment by mattn** [comment]
Some user want json1 but someone not.

**Comment by mattn** [comment]
If many users want json1 to be the default, then I will make it the default.

**Comment by otoolep** [comment]
Rebuilding the source code with the functionality you want, if it's not enabled by default, seems perfectly fine to me. I don't know why it's considered a big deal. It's the same with the SQLite source itself -- not everything is on by default. Thank you, @mattn and @rittneje for providing mattn/go-sqlite3 in the first place!

I recently did enable JSON support in [rqlite](https://github.com/rqlite/go-sqlite3/commit/41c9bb68d7fd3f531940911c9ad34a94493c4ae0), but only because the SQLite library that comes with Ubuntu has it enabled, so it seems mainstream enough. But still, building your own version is not a big deal IMHO.

**Comment by glebarez** [comment]
> Rebuilding the source code with the functionality you want, if it's not enabled by default, seems perfectly fine to me.

I agree with you here.
I think this whole thread has been drenched with misunderstanding 😄 

The behaviour I noticed on my side, that made me write a post here, is as follows.
I had to run every go command (go run / go test) with explicit "json1" build flag,
otherwise go would just rebuild this library with default settings, before executing, thus resulting in errors due to missing json functionality.
This behaviour can be clearly seen with -v flag in run or test.

I suspect it's rather go's behavioural issue.
